src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,569 @@
+                this.Container);
+            this.SetState(BatchWriterState.OperationCreated);
+
+            Debug.Assert(this.CurrentOperationContentId == null, "The Content-ID header is only supported in request messages.");
+
+            // write the operation's start boundary string
+            this.WriteStartBoundaryForOperation();
+
+            // write the headers and request separator line
+            ODataBatchWriterUtils.WriteResponsePreamble(this.RawOutputContext.TextWriter, changeSetBoundary != null, contentId);
+
+            return this.CurrentOperationResponseMessage;
+        }
+
+        /// <summary>
+        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// </summary>
+        protected override void StartBatchOperationContent()
+        {
+            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
+            Debug.Assert(this.RawOutputContext.TextWriter != null, "Must have a batch writer!");
+
+            // write the pending headers (if any)
+            this.WritePendingMessageData(false);
+
+            // flush the text writer to make sure all buffers of the text writer
+            // are flushed to the underlying async stream
+            this.RawOutputContext.TextWriter.Flush();
+        }
+
+        /// <summary>
+        /// Disposes the batch writer and set the 'OperationStreamRequested' batch writer state;
+        /// called after the flush operation(s) have completed.
+        /// </summary>
+        protected override void DisposeBatchWriterAndSetContentStreamRequestedState()
+        {
+            this.RawOutputContext.CloseWriter();
+
+            this.SetState(BatchWriterState.OperationStreamRequested);
+        }
+
+        /// <summary>
+        /// Verifies that calling CreateOperationRequestMessage is valid.
+        /// </summary>
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        protected override void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri, string contentId)
+        {
+            this.CanCreateOperationRequestMessageVerifierCommon(synchronousCall, method, uri, contentId);
+            VerifyCanCreateOperationRequestMessageAgainstChangeSetBoundary(method, contentId);
+        }
+
+        /// <summary>
+        /// Verifies that the writer is in correct state for the Flush operation.
+        /// </summary>
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        protected override void VerifyCanFlush(bool synchronousCall)

```

Comment:
>VerifyCanFlush [](start = 32, length = 14)

VerifyCanFlush should be a private method in the base class; see ODataWriterCore.  VerifyNotDisposed should be protected abstract and implemented in the derived class.

Reply:
This is related to note for VerifyCanFlush in base class.

---
In reply to: [141125904](https://github.com/OData/odata.net/pull/960#discussion_r141125904) [](ancestors = 141125904)

Reply:
See above comment for refactoring.

---
In reply to: [142579505](https://github.com/OData/odata.net/pull/960#discussion_r142579505) [](ancestors = 142579505,141125904)